zkevm-circuits/src/circuit_tools/memory.rs

```diff
@@ -159,46 +202,69 @@ impl<F: Field, C: CellType> MemoryBank<F, C> {
         self.cur.expr()
     }
 
-    pub(crate) fn load(
-        &self,
-        description: &'static str,
-        cb: &mut ConstraintBuilder<F, C>,
-        offset: Expression<F>,
-        values: &[Expression<F>],
-    ) {
-        self.load_with_key(description, cb, self.key() - offset, values);
+    fn query_write(&mut self) -> Cell<F> {
+        let cell = self.writes.0[self.writes.1].clone();
+        self.writes.1 += 1;
+        cell
     }
 
-    pub(crate) fn load_with_key(
-        &self,
-        description: &'static str,
-        cb: &mut ConstraintBuilder<F, C>,
-        key: Expression<F>,
-        values: &[Expression<F>],
-    ) {
-        cb.add_lookup(description, self.tag(), self.insert_key(key, values));
+    fn query_read(&mut self) -> Cell<F> {
+        let cell = self.reads.0[self.reads.1].clone();
+        self.reads.1 += 1;
+        cell
     }
 
     pub(crate) fn store(
-        &self,
+        &mut self,
         cb: &mut ConstraintBuilder<F, C>,
         values: &[Expression<F>],
     ) -> Expression<F> {
         let key = self.key() + 1.expr();
-        self.store_with_key(cb, key.expr(), values);
+        let condition = cb.get_condition_expr();
+        let values = self
+            .prepend_key(key.clone(), values)
+            .iter()
+            .map(|value| condition.expr() * value.expr())
+            .collect_vec();
+        let compressed_expr = cb.split_expression(
+            "compression",
+            rlc::expr(&values, cb.lookup_challenge.clone().unwrap().expr()),
+        );
+        let name = format!("{:?} write #{:?}", self.tag, self.writes.1);
+        cb.store_expression(
+            name.as_str(),
+            compressed_expr.expr(),
+            C::default(),
+            Some(self.query_write()),
+        );
+        self.table_conditions.push((cb.region_id, condition));
```

Comment:
Not possible to reuse the add_lookup of the `ConstraintBuilder`?

Reply:
I kinda just wanna leave the management of Memory to its own, if we go into `ConstraintBuilder` again we'll have to go into CM CellType and init the read/write columns within CM then later grab those columns for building lookup, that's why the last version was messy. 
But some code can be reused for sure, like I made an API on `circuit-tools-test` branch for the CB just to split expression with current condition.

Reply:
It basically does what `add_lookup` does, so should be fine? I want to prevent having lots of code that does the same thing and then having to make sure that all that code supports all the different features like degree reduction and regions, otherwise it's gonna get even more complicated.

Reply:
I still think this should reuse `add_lookup` like it did before, seems like that should be possible.

Reply:
Ok will do